virttest/qemu_devices/qcontainer.py

```diff
@@ -4031,9 +4031,41 @@ def hostdev_define_by_params(self, name, params, bus=None):
                 "failover_pair_id": params.get("vm_hostdev_failover_pair_id"),
             }
         )
+
+        devs = []
+
+        iommufd_id = params.objects("vm_hostdev_iommufd")
+        # To support single/multiple iommufd for single/multiple hostdev
+        for fd_id in iommufd_id:
+            if not self.get_by_qid(fd_id):
+                iommufd = self.iommufd_object_define_by_params(fd_id)
+                dev_params["iommufd"] = fd_id
+                devs.append(iommufd)
+                break
+            else:
+                # For multiple hostdev with the same iommufd
+                dev_params["iommufd"] = fd_id
+
         # TODO: Support vfio-ap and vfio-ccw, currently only for pci devices
         dev_bus = bus or {"aobject": params.get("pci_bus", "pci.0")}
         dev = qdevices.QDevice(driver, dev_params, parent_bus=dev_bus)
         for ext_k, ext_v in params.get_dict("vm_hostdev_extra_params").items():
             dev.set_param(ext_k, ext_v)
-        return dev
+        devs.append(dev)
+        return devs
+
+    def iommufd_object_define_by_params(self, obj_id):
+        """
+        Create iommufd object device by params
+
+        The iommufd objects cmdlines:
+            -object iommufd,id=iommufd0
+
+        :param obj_id: The id of the QObject device, e.g. iommufd0
+        :return: the iommufd QObject device
+        """
+        if not os.path.exists("/dev/iommu"):
+            raise ValueError("iommufd is not supported, modprobe iommufd to enable it")
```

Comment:
Hello @fanchen2, can you check creating an iommufd VM without `modprobe iommufd`? IIUC, qemu will automatically load it when you try to attend a iommufd to vfio. If this works, we can bypass this.

Sorry for my previous comment confusion here. But yes, modprobe in advance is good.

Reply:
@PaulYuuu 
iommufd cannot be removed directly, it has dependency with vfio.
To rmmod iommufd, we need to rmmod vfio_pci vfio_pci_core vfio_iommu_type1 vfio first. 
And when modprobe vfio(which is already be checked and loaded by avocado-vt), iommufd will be loaded together. So, we cannot try with vfio loaded but without iommufd loaded. 
If modprobe iommufd with vfio, iommufd can be loaded, and /dev/iommu can be found, but vfio will not be loaded together, the case cannot pass.
So in this case, looks like we can skip this check? What di you think? If so, let me remove it.

Reply:
Sounds good, thank you.